library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

```diff
@@ -0,0 +1,1208 @@
+package com.otaliastudios.zoom
+
+import android.content.Context
+import android.graphics.Matrix
+import android.graphics.RectF
+import android.os.Build
+import android.support.annotation.IntDef
+import android.view.*
+import android.view.animation.AccelerateDecelerateInterpolator
+import android.widget.OverScroller
+import com.otaliastudios.zoom.ZoomApi.*
+
+
+/**
+ * A low level class that listens to touch events and posts zoom and pan updates.
+ * The most useful output is a [Matrix] that can be used to do pretty much everything,
+ * from canvas drawing to View hierarchies translations.
+ *
+ *
+ * Users are required to:
+ * - Pass the container view in the constructor
+ * - Notify the helper of the content size, using [.setContentSize]
+ * - Pass touch events to [.onInterceptTouchEvent] and [.onTouchEvent]
+ *
+ *
+ * This class will apply a base transformation to the content, see [.setTransformation],
+ * so that it is laid out initially as we wish.
+ *
+ *
+ * When the scaling makes the content smaller than our viewport, the engine will always try
+ * to keep the content centered.
+ */
+class ZoomEngine
+/**
+ * Constructs an helper instance.
+ *
+ * @param context   a valid context
+ * @param container the view hosting the zoomable content
+ * @param listener a listener for events
+ */
+constructor(context: Context, private var container: View, var listener: Listener?)
```

Comment:
Can we do it the other way round?

Primary constructor should be `internal constructor(context: Context)`.

Then a secondary constructor will be `public constructor(context: Context, container: View, var listener: Listener?)`. This secondary constructor does two things, calls `this(context)` and calls `setContainer(container)` function.

About the setContainer function, it should be called only once and be `internal`. Something like

```kotlin
private lateinit var mView: View

internal fun setContainer(container: View) {
    mView = container
    // Register onGlobalLayout
}
```

This is to

- delegate constructors the correct way, avoiding object: View(context) {}
- internally keep the `mView` name for the variable, which is more readable than `container` since we also have `mViewWidth` `mViewHeight` and all m* stuff.
- avoid calling addOnGlobalLayoutListener(this) twice (both in the init block and in setContainer function)

Reply:
done